primer/src/Primer/Eval/Redex.hs

```diff
@@ -250,6 +250,11 @@ data Redex
     --  nor any free variable of @e@ (to avoid capture)]
     -- [We actually do this rule for a whole sequence of let bindings at once]
     -- [If we push into an annotation, we drop term variables:  let x = e in (t : T)  ~> (let x = e in t) : T]
+    -- [We actually drop all "pointless" variables:
+    --   let x = e in f s  ~>  (let x = e in f) (let x = e in s)    if @x@ is free in @f@, and free in @s@
+    --   let x = e in f s  ~>  f (let x = e in s)                   if @x@ is not free in @f@, but free in @s@
+    --   let x = e in f s  ~>  (let x = e in f) s                   if @x@ is free in @f@, but not free in @s@
+    --   let x = e in f s  ~>  f s                                  if @x@ is not free in @f@, and not free in @s@
```

Comment:
This does somewhat violate our "eval steps should be as small/simple as possible" maxim. I think it's worth it here though. We'll just need to be careful about how we explain it in the UI.

Reply:
Per discussion in our 2023-07-25 developer meeting:

* We do this in the full evaluator for performance reasons, to prevent evaluation of a pointless push down
* There is some concern that in step eval, it will not be clear why the `let` isn't pushed down into both subtrees
* Therefore in step eval, pushing down a `let` will push down into both subtrees

This marks a branch in the evaluator's development: up to this point, the full and step evaluators have shared the same implementation. However, we already know that they will need to diverge in the near future so that we can start making optimizations in the full evaluator (its performance is too slow to be practical), so this divergence was inevitable anyway.

@georgefst points out that this may be ... (will return to this comment in a bit)

Reply:
In a meeting today we decided:
- [ ] I'll document the performance impact
- [ ] We will make it optional, so full eval will push+elide but step eval will simply push (and elide in a separate step)

We also discussed the similar violation where we simultaneously push multiple adjacent `let`s, and for those I will also
- [ ] Document performance impact
- [ ] Document the fact we simultaneously push (it is not super obvious!) -- probably write a note and link from the option definition and also the implementation
- [ ] make it optional, in same way as above

Note that we have two performance-impacting things, and historically the push+elide was implemented first, to fix the perf regression in `unit_8`. These two obviously interact, and I am not actually sure how much of the impact comes from each -- I should probably get 4 perf numbers (4 = 2*2 option settings).

Reply:
One interesting unintentional interaction is that non-simultaneous-push is actually more aggressive with eliding redundant lets. In `let a = _ in let b = _ in λc. b` with simultaneous-push we will push both `let`s under the lambda. But with non-simultaneous-push we will consider whether 'let a' is a redex and decide it is not a `PushLet` redex as there is a stack of lets underneath it, but it is an `ElideLet` redex (as we fall through to the next case of `viewRedex`). This is an inconsistency that gives extremely misleading performance numbers.

Now that I understand what is happening, I can obtain sensible measurements.

Reply:
Some timing data follows.

I ran `mapEven` for various options and lengths of lists (this is roughly `unit_8`), and also the "worker/wrapper"/`unit_9` variation. This was done with a very simple script, so the numbers are probably somewhat noisy.
(Note that we have 3 options now: the two mentioned above, and "aggressiveElision" which prefers to elide early, rather tan pushing all the way to a leaf before eliding)

For `n=5`, I get 
pushMulti  |  aggressiveElision  |  pushAndElide | map, steps | map, time (s) | map/ww, steps | map/ww, time (s)
-----------|---------------------|---------------|------------|---------------|---------------|-----------------
True | True | True | 450 | 0.129396629000 | 368 | 0.130995374000
False | True | True | 450 | 0.118675216000 | 373 | 0.111896999000
True | False | True | 450 | 0.130649372000 | 368 | 0.131270030000
False | False | True | 450 | 0.116904985000 | 373 | 0.110885103000
True | True | False | 583 | 0.170570243000 | 461 | 0.162431173000
False | True | False | 628 | 0.169462742000 | 518 | 0.157327909000
True | False | False | 687 | 1.926900242000 | 500 | 0.440785217000
False | False | False | 997 | 2.959247222000 | 1465 | 4.398923352000

And for `n=10` I  get
pushMulti  |  aggressiveElision  |  pushAndElide | map, steps | map, time (s) | map/ww, steps | map/ww, time (s)                                                                                                                                                                                                               
-----------|---------------------|---------------|------------|---------------|---------------|-----------------                                                                                                                                                                                                               
True | True | True | 1145 | 0.389490347000 | 923 | 0.357104768000                                                                                                                                                                                                                                                              
False | True | True | 1145 | 0.364744574000 | 933 | 0.320479639000                                                                                                                                                                                                                                                             
True | False | True | 1145 | 0.385016614000 | 923 | 0.358131104000                                                                                                                                                                                                                                                             
False | False | True | 1145 | 0.363208693000 | 933 | 0.316983115000                                                                                                                                                                                                                                                            
True | True | False | 1478 | 0.540102081000 | 1151 | 0.455796149000                                                                                                                                                                                                                                                            
False | True | False | 1588 | 0.566575896000 | 1258 | 0.459367898000                                                                                                                                                                                                                                                           
True | False | False | 1707 | 518.533781665000 | 1250 | 4.326406590000 

(The `False,False,False` configuration` took over 2000 steps)

Reply:
So my brief analysis is:

* `map/ww` is between 39% to 85% worse performance with push-down `let`s, in the best case scenario (F/F/T). It is still faster than straightforward `map`, however, except for a few odd cases with `n = 5` where I assume there is just too much noise.
* `map` is between to  14% to 25% worse performance with push-down `let`s, in the best case scenario (also F/F/T).
* The performance hit of push-down `let`s decreases as `n` increases.
* For these benchmarks, at least, enabling `pushAndElide` and disabling the other optimizations (i.e., F/F/T) is the winner.
* For these benchmarks, enabling `pushMulti` appears to always be worse, and when only that feature is enabled, it is several orders of magnitude worse on the `map` benchmark, which is weird. @brprice do you have a handle on why this might be?

This is a difficult call. My intuition is that pushing down `let`s is going to be a good pedagogical feature, for the reasons explained in #28, but assuming that the performance impact seen on this benchmark is indicative, push-down `let`s will come with a not insignificant performance cost, at a time when we're particularly concerned about improving performance.

@brprice would it be feasible to keep around the current behavior as yet another option? Ideally we could keep it around for both full- and step-evaluation, but if it were easier only to keep it around for full eval, that would also be OK, I think, and fairly easily explainable as an optimization that the full evaluator can perform.

Reply:
> The performance hit of push-down `let`s decreases as `n` increases.

I think it might be useful to see a comparison of the old behaviour to the new one with all optimisations, on larger programs. Then we can check whether these continue to converge.

Reply:
That's an interesting point. It could be that once we've optimized the full evaluator that, for large enough `n` such that the runtime is still noticeable, the performance impact of push-down `let`s is insignificant.

Reply:
Some new numbers (I fixed a silly bug...)

Running the whole testsuite:
      | no PBTs (via hyperfine) | with PBTs (three manual runs, sorted)
old | 482.1 ms ±   8.5 ms | 73.4s, 76.6s, 85.6s
new | 576.6 ms ±  29.9 ms | 85.9s, 106.6s, 116.3s

I'll edit some longer runs in here shortly...

Reply:
OK, so to make sure I understand correctly, pretty much all the analysis I did earlier is invalid? :)

Reply:
I'm afraid so...

Last batch of numbers: memory usage. This one is very in favor of the new approach

With `n=100`:
|| Total alloc bytes | copied during gc bytes | max residency bytes | total memory in use | productivity
|---|---|---|---|---|---
map, this PR | 251,370,361,304 | 45,984,704,024 | 11,501,960 | 52 MiB | 59.2%
map, orig | 192,034,025,032 | 35,501,492,712 | 400,189,696 |  1112 MiB |  55.6%
map/ww, this PR |206,965,550,480| 41,429,113,248|12,180,304|52 MiB|58.0%
map/ww, orig |168,011,848,872|33,181,823,024|358,839,832|994 MiB| 54.1%

Reply:
> This one is very in favor of the new approach

That's a pretty dramatic difference - any idea why?

Reply:
I was expecting some difference in favor of the new approach because we simply have to remember less when reducing. Since our rules are now local, we don't need to carry around a context of all (distantly) enclosing `let`s and which ones can be substituted without causing capture. 

However, I am surprised the difference is this large.

Reply:
In our meeting today, we decided to go ahead with this PR regardless of the performance impact. Our reasoning was that 
- we are aiming for pedagogy of eval, rather than performance
- the much lower memory footprint may (perhaps) be more important to overall performance than the speed (for memory-constrained systems)
- the "full evaluator" is a simple wrapper around the interactive/step evaluator -- if we want to improve performance, we should change strategy (eval on a virtual machine; do some compilation; etc), though it is not clear if/how this could be integrated into the step eval as a "jump forward"/"don't show me steps inside that subexpression" (due to (perhaps) wanting to stop before reaching a normal form and needing to reify a result back to a term; also due to (perhaps) wanting to evaluate open terms (c.f. #1055))

Reply:
When rebasing this PR, I have found that the above memory impact was actually an orthogonal change, which is now contained in #1110. I will generate some new numbers shortly.

Reply:
New numbers, for both before and after this PR. Both are based off the new strictness annotation, #1110:

Memory, generated by
- `unit tests`: `cabal run -O2 primer-test -- --hide-successes --hedgehog-tests 0 +RTS -s` (only for the default set of options, as there are some tests that require such)
- `big 8, n=...`: `cabal run -O2 primer-test -- --hide-successes -p EvalFull.8 +RTS -s`, after editing `EvalFull.unit_8` to change `n` and bump the upper bound from 2000 to 200000 and `EvalFull.evalFullTest` to set the options

| Branch  | Test suite | push Multi | aggr. Elision | push And Elide | alloc heap (B) | copy gc (B) | max res (B) | max slop (B) | total mem (MB) |
|:--------|:-----------|----------:|------------------:|-------------:|-----------:|--------:|--------:|---------:|----------:|
| main    | unit tests | N/A  | N/A  | N/A  | 2,253,549,080 | 81,112,304 | 5,381,576 | 245,432 | 34 | 
| this pr | unit tests | True | True | True | 2,254,933,496 | 85,541,920 | 4,884,232 | 120,912 | 33 | 
| main    | big 8, 10  | N/A  | N/A  | N/A  | 687,433,096 | 14,499,296 | 1,679,576 | 61,224 | 23 | 
| this pr | big 8, 10  | True | True | True | 667,293,888 | 23,986,928 | 1,806,288 | 76,024 | 24 | 
| this pr | big 8, 10  | True | True | False| 951,432,832 | 54,166,384 | 2,183,864 | 75,792 | 24 | 
| this pr | big 8, 10  | True | False| True | 667,293,912 | 24,121,216 | 1,835,000 | 67,728 | 24 | 
| this pr | big 8, 10  | False| True | True | 726,874,408 | 20,025,288 | 1,807,416 | 76,744 | 25 | 
| this pr | big 8, 10  | False| True | False| 1,060,585,984 | 41,553,904 | 1,913,520 | 74,824 | 25 | 
| this pr | big 8, 10  | False| False| True | 723,390,560 | 20,034,168 | 1,806,304 | 69,664 | 25 | 
| main    | big 8, 20  | N/A  | N/A  | N/A  | 2,590,407,608 | 73,241,792 | 2,105,768 | 73,304 | 24 | 
| this pr | big 8, 20  | True | True | True | 2,581,434,568 | 126,270,680 | 2,276,296 | 81,544 | 26 | 
| this pr | big 8, 20  | True | True | False| 3,920,475,240 | 334,034,784 | 3,275,976 | 95,000 | 28 | 
| this pr | big 8, 20  | True | False| True | 2,581,434,584 | 127,493,016 | 2,181,192 | 80,168 | 25 | 
| this pr | big 8, 20  | False| True | True | 2,771,095,368 | 108,475,248 | 2,264,312 | 144,136 | 26 | 
| this pr | big 8, 20  | False| True | False| 4,363,811,480 | 294,687,904 | 3,085,192 | 128,832 | 28 | 
| this pr | big 8, 20  | False| False| True | 2,762,316,576 | 108,085,664 | 2,202,808 | 75,832 | 26 | 
| main    | big 8, 50  | N/A  | N/A  | N/A  | 27,423,073,040 | 2,424,616,608 | 4,649,376 | 496,840 | 32 | 
| this pr | big 8, 50  | True | True | True | 27,014,099,816 | 3,321,942,792 | 5,212,016 | 551,048 | 34 | 
| this pr | big 8, 50  | True | True | False| 40,758,062,920 | 7,319,097,552 | 9,713,800 | 672,720 | 46 | 
| this pr | big 8, 50  | True | False| True | 27,014,128,744 | 3,301,876,928 | 5,222,072 | 552,240 | 37 | 
| this pr | big 8, 50  | False| True | True | 28,017,813,904 | 3,098,616,752 | 5,175,104 | 530,616 | 35 | 
| this pr | big 8, 50  | False| True | False| 44,494,569,304 | 7,371,045,888 | 9,545,024 | 604,440 | 45 | 
| this pr | big 8, 50  | False| False| True | 27,964,518,976 | 3,117,906,312 | 5,248,384 | 546,112 | 36 | 
| main    | big 8, 100 | N/A  | N/A  | N/A  | 195,684,938,608 | 34,688,595,528 | 13,795,496 | 1,632,720 | 60 | 
| this pr | big 8, 100 | True | True | True | 190,533,645,264 | 43,294,432,904 | 15,906,656 | 1,651,328 | 68 | 
| this pr | big 8, 100 | True | True | False| 283,194,326,656 | 89,053,479,520 | 33,821,448 | 2,200,456 | 114 | 
| this pr | big 8, 100 | True | False| True | 190,533,646,264 | 43,274,133,832 | 17,822,504 | 1,760,472 | 70 | 
| this pr | big 8, 100 | False| True | True | 194,584,842,160 | 42,208,027,712 | 17,695,144 | 1,715,800 | 70 | 
| this pr | big 8, 100 | False| True | False| 306,139,099,160 | 93,087,751,416 | 32,043,520 | 1,914,912 | 111 | 
| this pr | big 8, 100 | False| False| True | 194,292,871,256 | 42,235,823,416 | 17,311,496 | 1,707,968 | 70 | 


Time, via `hyperfine "$(cabal list-bin -O2 primer-test) --..."` where the time is short enough to get sensible results, but just one run for large `n`

| Branch  | Test suite | push Multi | aggr. Elision | push And Elide | Mean | Min | Max |
|:--------|:-----------|----------:|------------------:|-------------:|-----:|----:|----:|
| main    | unit tests | N/A   | N/A   | N/A   | 447.7 ± 9.1 ms | 425.8 ms | 458.2 ms |
| this pr | unit tests | True  | True  | True  | 470.4 ± 13.6 ms | 445.6 ms | 485.4 ms |
| main    | big 8, 10  | N/A   | N/A   | N/A   | 276.5 ± 4.2 ms | 273.9 ms | 284.3 ms |
| this pr | big 8, 10  | True  | True  | True  | 284.6 ± 0.7 ms | 284.1 ms | 286.6 ms |
| this pr | big 8, 10  | True  | True  | False | 408.7 ± 5.3 ms | 404.1 ms | 415.8 ms |
| this pr | big 8, 10  | True  | False | True  | 286.4 ± 3.1 ms | 284.3 ms | 294.8 ms |
| this pr | big 8, 10  | False | True  | True  | 305.8 ± 3.1 ms | 304.1 ms | 314.1 ms |
| this pr | big 8, 10  | False | True  | False | 439.7 ± 7.2 ms | 434.0 ms | 454.8 ms |
| this pr | big 8, 10  | False | False | True  | 298.4 ± 5.1 ms | 294.0 ms | 304.3 ms |
| main    | big 8, 20  | N/A   | N/A   | N/A   | 953.4 ± 9.6 ms | 934.2 ms | 966.8 ms |
| this pr | big 8, 20  | True  | True  | True  |  1.012 ± 0.007 s | 1.004 s | 1.025 s |
| this pr | big 8, 20  | True  | True  | False |  1.626 ± 0.010 s | 1.605 s | 1.635 s |
| this pr | big 8, 20  | True  | False | True  |  1.026 ± 0.006 s | 1.016 s | 1.037 s |
| this pr | big 8, 20  | False | True  | True  |  1.063 ± 0.008 s | 1.054 s | 1.075 s |
| this pr | big 8, 20  | False | True  | False |  1.744 ± 0.009 s | 1.725 s | 1.755 s |
| this pr | big 8, 20  | False | False | True  |  1.055 ± 0.007 s | 1.044 s | 1.065 s |
| main    | big 8, 50  | N/A   | N/A   | N/A   | 10.858 ± 0.062 s | 10.785 s | 10.985 s |
| this pr | big 8, 50  | True  | True  | True  | 11.320 ± 0.053 s | 11.238 s | 11.405 s |
| this pr | big 8, 50  | True  | True  | False | 18.695 ± 0.129 s | 18.499 s | 18.949 s |
| this pr | big 8, 50  | True  | False | True  | 11.332 ± 0.075 s | 11.255 s | 11.475 s |
| this pr | big 8, 50  | False | True  | True  | 11.602 ± 0.052 s | 11.495 s | 11.685 s |
| this pr | big 8, 50  | False | True  | False | 20.096 ± 0.119 s | 19.907 s | 20.228 s |
| this pr | big 8, 50  | False | False | True  | 11.474 ± 0.078 s | 11.366 s | 11.568 s |
| main    | big 8, 100 | N/A  | N/A  | N/A  | 102.97 s (one run) | N/A | N/A |
| this pr | big 8, 100 | True  | True  | True  | 97.34s (one run) | N/A | N/A |
| this pr | big 8, 100 | True  | True  | False | 160.41s (one run) | N/A | N/A |
| this pr | big 8, 100 | True  | False | True  | 97.33s (one run) | N/A | N/A |
| this pr | big 8, 100 | False | True  | True  | 100.89s (one run) | N/A | N/A |
| this pr | big 8, 100 | False | True  | False | 170.40s (one run) | N/A | N/A |
| this pr | big 8, 100 | False | False | True  | 99.44s (one run) | N/A | N/A |

Reply:
Yay! So unless I'm missing something, it looks like with the strictness annotation fix in #1110, and with all 3 push-down `let` optimizations enabled, this PR is now more or less a wash for both wall-clock performance and memory allocations, at least for the benchmarks we've considered here. Is that correct?

Reply:
Yeah, more-or-less. Which agrees with my initial assessment (phew!)

Reply:
That's great news!